splitter: ingest+split logs from CWLogs subscription

nathanleiby opened the pull request:
Using this to test out new functionality. We may end up moving this to a separate service or lambda, to split logs before they get to the consumer.

nathanleiby committed:
splitter: ingest+split logs from CWLogs subscription

nathanleiby committed:
splitter: docs and linting

xavi- reviewed (approved):
Lgtm.  Is there an example of how'd it be used by a consumer?

nathanleiby commented:
@xavi- not yet, planning on testing today and making an example

nathanleiby committed:
writer: add MockFirehoseAPI for tests

generated via

```
mockgen github.com/aws/aws-sdk-go/service/firehose/firehoseiface FirehoseAPI > mock_firehose_api_test.go
```

nathanleiby committed:
splitter: refactor

nathanleiby committed:
writer: process records from CWLogs subscription

also adds tests

nathanleiby committed:
splitter: make Split public, for use in decode tests

nathanleiby committed:
linting fixes

- move mocks
- ignore mocks in Makefile's PKGS, so we don't fail lint on auto-generated files
- lint fix for Splitter

nathanleiby committed:
glide: update deps -- add gomock

nathanleiby committed:
glide: update aws-sdk-go version

nathanleiby committed:
splitter: allow filtering logs by env

Currently, we don't have the capability to either:

1. control the log group that AWS Batch writes to (everything goes to a single log group, dev or prod)
2. filter by log stream name within a CWLogs subscription

We've asked AWS to fix (1), and in the meantime, are doing a solution
similar to (2) but in the consumer's code.

nathanleiby commented:
@xavi- Hey! Would you mind taking another pass on this one (when time permits), given all the changes since you original review?

In its current state, it should be

- backwards compatible
- able to handle a CWLogs subscription being added to a Kinesis stream

**Rollout Plan:**

I plan to deploy in Dev and monitor there for a bit first to make sure everything looks good. Then I'll enable a CW Logs subscription to the Dev kinesis stream, and verify that workflow logs are parsed correctly and arrive in Kibana / Archive.

After this, I'll deploy in prod.

nathanleiby commented:
@xavi- One other note: this hasn't yet solved the problem of doing better checkpointing on Batch messages. Currently, if any message from the CW Logs record is checkpointed, it will assume the record was successful.

https://github.com/Clever/kinesis-to-firehose/blob/c7fc93a7b528776d68bb2d0ce8da44d1d130e36f/writer/firehose_writer.go#L160-L161

nathanleiby commented:
I'm going to think on the right solution for that one, but I don't think it blocks deploying this in dev and being sure it works correctly.

nathanleiby commented:
@xavi- there's one other workaround here that @mohit and I discussed offline. Given the way subscriptions work, it's hard to target dev logs to dev stream and prod logs to prod stream, so here's the workaround we came up with: https://github.com/Clever/kinesis-to-firehose/pull/21/commits/c7fc93a7b528776d68bb2d0ce8da44d1d130e36f

xavi- commented on writer/firehose_writer.go:
```diff
@@ -136,14 +137,45 @@ func (f *FirehoseWriter) ProcessRecords(records []kcl.Record) error {
+	}
+
+	// Process a single message, from KPL
+	return f.parseMessageAndPrepareToSend(data, record.SequenceNumber, record.SubSequenceNumber)
```
Nit. Perhaps reverse conditional here to:

```golang
if !splitter.IsGzipped(data) { // Process single message...
    // ...
} else { // Process messages from CWLogs
    // ...
    // ...
}
```

nathanleiby committed:
splitter: (nit) cleanup extra comment

xavi- commented on splitter/splitter.go:
```diff
@@ -0,0 +1,114 @@
+func GetMessagesFromGzippedInput(input string, prodEnv bool) ([]string, error) {
+	unpacked, err := unpack(input)
+	if err != nil {
+		fmt.Println("ERROR: " + err.Error())
```
Debug statement?

nathanleiby committed:
splitter/writer: pr feedback
